src/core/reference/include/openvino/reference/equal.hpp

```diff
@@ -36,14 +35,46 @@ void equal(const T* arg0, const T* arg1, char* out, size_t count) {
  * @param arg_shape1      Input 1 shape.
  * @param broadcast_spec  Broadcast specification mode.
  */
-template <typename T, typename U, typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+template <typename T, typename U, typename std::enable_if_t<std::is_integral_v<T>>* = nullptr>
 void equal(const T* arg0,
            const T* arg1,
            U* out,
            const Shape& arg0_shape,
            const Shape& arg1_shape,
            const op::AutoBroadcastSpec& broadcast_spec) {
-    autobroadcast_binop(arg0, arg1, out, arg0_shape, arg1_shape, broadcast_spec, func::equal<T>);
+    if constexpr (sizeof(T) == 1) {
+        autobroadcast_binop(reinterpret_cast<const int8_t*>(arg0),
+                            reinterpret_cast<const int8_t*>(arg1),
+                            out,
+                            arg0_shape,
+                            arg1_shape,
+                            broadcast_spec,
+                            func::equal<int8_t>);
+    } else if constexpr (sizeof(T) == 2) {
+        autobroadcast_binop(reinterpret_cast<const int16_t*>(arg0),
+                            reinterpret_cast<const int16_t*>(arg1),
+                            out,
+                            arg0_shape,
+                            arg1_shape,
+                            broadcast_spec,
+                            func::equal<int16_t>);
+    } else if constexpr (sizeof(T) == 4) {
+        autobroadcast_binop(reinterpret_cast<const int32_t*>(arg0),
+                            reinterpret_cast<const int32_t*>(arg1),
+                            out,
+                            arg0_shape,
+                            arg1_shape,
+                            broadcast_spec,
+                            func::equal<int32_t>);
+    } else if constexpr (sizeof(T) == 8) {
+        autobroadcast_binop(reinterpret_cast<const int64_t*>(arg0),
+                            reinterpret_cast<const int64_t*>(arg1),
+                            out,
+                            arg0_shape,
+                            arg1_shape,
+                            broadcast_spec,
+                            func::equal<int64_t>);
+    }
```

Comment:
Of course the general goal is to decrease binary size, and the effort to check this approach is reasonable, but considering the actual gain here. Is it worth to introduce additional layer of type inference logic to save ~12KB? 

cc: @praasz

Reply:
Code bloat was removed .. it's even less code than before.